common/network-common/src/main/java/org/apache/spark/network/protocol/MessageWithHeader.java

```diff
@@ -137,30 +137,15 @@ protected void deallocate() {
   }
 
   private int copyByteBuf(ByteBuf buf, WritableByteChannel target) throws IOException {
-    ByteBuffer buffer = buf.nioBuffer();
-    int written = (buffer.remaining() <= NIO_BUFFER_LIMIT) ?
-      target.write(buffer) : writeNioBuffer(target, buffer);
+    // SPARK-24578: cap the sub-region's size of returned nio buffer to improve the performance
+    // for the case that the passed-in buffer has too many components.
+    int length = Math.min(buf.readableBytes(), NIO_BUFFER_LIMIT);
+    ByteBuffer buffer = buf.nioBuffer(buf.readerIndex(), length);
```

Comment:
I think you can go one step further here, and call `buf.nioBuffers(int, int)` (plural) 
https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L2355

that will avoid the copying required to create the merged buffer (though its a bit complicated as you have to check for incomplete writes from any single `target.write()` call).

Also OK to leave this for now as this is a pretty important fix.

Reply:
Sure, I will make a follow up PR to address this.

Reply:
Why not do this in this PR since this is a small change and we don't have a new release recently?

Reply:
this pr is fixing a pretty serious issue, we know Wenbo is going to roll this out immediately, and I suspect even more users will.  this fix is also "obviously correct" -- the followup here is not super complicated, but also will be more prone to bugs.  So I'm inclined to just get this in.

anyway if @WenboZhao can do the other part today, then sure, but I think we should get this in quickly.

Reply:
Thanks @squito and @zsxwing. I would prefer to do it in a different PR with more careful benchmark and testing. As @squito, that change is more prone to bugs.

Reply:
@WenboZhao did you ever follow up on this, or at least file another jira for it?  sorry if I missed it

Reply:
for anybody watching this, eventually SPARK-25115 was opened (currently has a PR)